app/components/channel_loader/channel_loader.js

```diff
@@ -9,15 +9,79 @@ import {
 } from 'react-native';
 import Placeholder from 'rn-placeholder';
 
+import EventEmitter from 'mattermost-redux/utils/event_emitter';
+
+import CustomPropTypes from 'app/constants/custom_prop_types';
 import {changeOpacity, makeStyleSheetFromTheme} from 'app/utils/theme';
 
 export default class ChannelLoader extends PureComponent {
     static propTypes = {
+        actions: PropTypes.shape({
+            handleSelectChannel: PropTypes.func.isRequired,
+            markChannelAsViewed: PropTypes.func.isRequired,
+            markChannelAsRead: PropTypes.func.isRequired,
+            setChannelLoading: PropTypes.func.isRequired,
+        }).isRequired,
         backgroundColor: PropTypes.string,
         channelIsLoading: PropTypes.bool.isRequired,
+        maxRows: PropTypes.number,
+        style: CustomPropTypes.Style,
         theme: PropTypes.object.isRequired,
     };
 
+    static defaultProps = {
+        maxRows: 6,
+    };
+
+    state = {
+        switching: false,
+    };
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        if (!nextProps.channelIsLoading && prevState.switch) {
+            return {
+                switch: false,
+                channel: null,
+                currentChannelId: null,
+            };
+        }
+
+        return prevState;
+    }
+
+    componentDidMount() {
+        EventEmitter.on('switch_channel', this.handleChannelSwitch);
```

Comment:
Why use the event emitter for this? Couldn't we detect this by watching if the current channel ID changes?

Reply:
yes we could but it was taking longer to execute than doing this, so this is for performance really